### ISSUE TITLE
List request timeout not respected during response write stage

### ISSUE THREAD

**Issue by natherz97** [kubernetes/kubernetes#119415]
### What happened?

It is possible to issue a list request which exceeds the duration set by the request-timeout apiserver flag. The request-timeout flag does not appear to be enforced when the apiserver is serializing the response and writing it back to the client (whereas it is enforced on other request stages such as when listing from storage). Specifically, slow clients with insufficient CPU or network resources, can introduce arbitrarily high list latencies by being slow to receive the response. These requests will occupy APF seats for the entire duration 1+ minute duration they are executing which can starve other clients using the same FlowSchema. 

### What did you expect to happen?

We expected the list requests to be subject to the duration specified by the request-timeout flag.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create 50k secrets.
2. List all the secrets in this namespace and force the HTTP client to be slow in receiving the response by leveraging the limit-rate flag.
```
curl --cert ./client.pem --key ./client-key.pem --cacert ./ca.pem $server/api/v1/namespaces/default/secrets --limit-rate 200k | wc -c
```
3. APIServer logs show the request took 5+ minutes to complete:
```
I0714 19:37:10.706663 11 trace.go:205] Trace[449374747]: "List" url:/api/v1/namespaces/default/secrets,user-agent:curl/8.0.1,audit-id:a09c55c2-0f84-4d6e-8fe6-fcde2e226b31,client:205.251.233.233,accept:*/*,protocol:HTTP/2.0 (14-Jul-2023 19:31:34.658) (total time: 336048ms):
Trace[449374747]: ---"Listing from storage done" 13749ms (19:31:48.408)
Trace[449374747]: ---"Writing http response done" count:40803 322298ms (19:37:10.706)
Trace[449374747]: [5m36.04813648s] [5m36.04813648s] END
```
4. The audit log entry gives the latency breakdown:
```
  "annotations": {
    "apiserver.latency.k8s.io/etcd": "13.522549999s",
    "apiserver.latency.k8s.io/response-write": "5m21.445820827s",
    "apiserver.latency.k8s.io/serialize-response-object": "5m22.298206407s",
    "apiserver.latency.k8s.io/total": "5m36.049590964s",
    "apiserver.latency.k8s.io/transform-response-object": "1.724µs",
    "authorization.k8s.io/decision": "allow",
    "authorization.k8s.io/reason": "RBAC: allowed by ClusterRoleBinding \"eks:k8s-metrics\" of ClusterRole \"eks:k8s-metrics\" to User \"eks:k8s-metrics\""
  }
```
5. Checking APF concurrency in use shows the global-default bucket is using 10 seats for the full 5 minute request:
```
 kubectl get --raw /metrics --kubeconfig kubeconfig | grep apiserver_flowcontrol_request_concurrency_in_use
...
apiserver_flowcontrol_request_concurrency_in_use{flow_schema="global-default",priority_level="global-default"} 10
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.13-eks-0a21954", GitCommit:"6305d65c340554ad8b4d7a5f21391c9fa34932cb", GitTreeState:"clean", BuildDate:"2023-04-15T00:37:31Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.13-eks-0a21954", GitCommit:"6305d65c340554ad8b4d7a5f21391c9fa34932cb", GitTreeState:"clean", BuildDate:"2023-04-15T00:33:45Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"
```

</details>

### Cloud provider

<details>
AWS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by shyamjvs** [comment]
/sig scalability

cc @wojtek-t @andrewsykim @MikeSpreitzer

**Comment by wojtek-t** [comment]
The main problem is that writing the response doesn't really conform to timeouts, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L134

In the ideal world, we would like io.Writer interface to accept context, but in the meantime we may want to workaround it somethow.

cc @mborsz

**Comment by natherz97** [comment]
A related issue is if the client is slow to receive the init events from a newly established watch. In this case, all seats from a given priority level can be exhausted by a large enough number of watch requests each holding 1 seat indefinitely. The case mentioned in this issue with list would be covered by passing a context to a writer, but since watch doesn't have a timeout enforced, this case wouldn't be covered.

There was a previous discussion related to watch requests holding seats indefinitely against an uninitialized watch-cache where @shyamjvs suggested that the server should enforce a timeout on the initial phase of the watch request: https://github.com/kubernetes/kubernetes/pull/115656#issuecomment-1439051697. @wojtek-t do you have any thoughts on this case with watch or should this be tracked as a separate issue?

**Comment by natherz97** [comment]
One more form of this issue which would impact both watch or list requests (in addition to low CPU/network bandwidth) is if the client process is paused/dead locked when the apiserver is writing the HTTP response back to the client. Even if the client is setting a client-side timeout to workaround the server-side timeout not being respected, the client wouldn't be running to terminate the request on its side. The client thread would have to be resumed to receive the rest of the response/terminate the request or the underlying TCP connection would need to be terminated. 

Here's a repro of this form of the issue with pausing the kubectl process when it was receiving the response:
```
# Issue a list pods request with a client-side timeout of 5s and then pause the process:
% time kubectl get --raw "/api/v1/namespaces/default/pods" --request-timeout 5s | wc -c
^Z
zsh: suspended  kubectl get --raw "/api/v1/namespaces/default/pods" --request-timeout 5s | wc -c

# Wait 30 minutes and then resume the process so that the list pod request completes:
% fg
[19]  + continued  kubectl get --raw "/api/v1/namespaces/default/pods" --request-timeout 5s | wc -c
Unable to connect to the server: context deadline exceeded (Client.Timeout exceeded while awaiting headers)
0
kubectl get --raw "/api/v1/namespaces/default/pods" --request-timeout 5s  0.62s user 0.08s system 0% cpu 32:02.06 total
wc -c  0.00s user 0.00s system 0% cpu 32:02.06 total
```
Checking audit log latency annotations shows the request was executing and holding APF seats for 30 minutes:
```
  "annotations": {
    "apiserver.latency.k8s.io/etcd": "619.47221ms",
    "apiserver.latency.k8s.io/response-write": "32m4.915489787s",
    "apiserver.latency.k8s.io/serialize-response-object": "32m6.987983593s",
    "apiserver.latency.k8s.io/total": "32m8.894443448s",
    "apiserver.latency.k8s.io/transform-response-object": "2.08µs",
    "authorization.k8s.io/decision": "allow"
  }
```

**Comment by wojtek-t** [comment]
We discussed this on last SIG meeting - here is the link to notes:
https://docs.google.com/document/d/1hEpf25qifVWztaeZPFmjNiJvPo-5JX1z0LSvvVY5G2g/edit#bookmark=id.oh2c71gk85ww

The first part of the mitigation is https://github.com/kubernetes/kubernetes/pull/114189 - we will keep thinking about how we can further mitigate the risk here.

@tkashem - FYI

**Comment by natherz97** [comment]
I was digging more how a slow client would impact watch requests and it doesn't look like a slow client receiving InitEvents would increase the time the 1 seat is held based on this comment: https://github.com/kubernetes/kubernetes/blob/1f3256bbac668f120433990028ca1fe206e4117b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L510-L516

Also mentioned in APF KEP, "In watchcache, the initialization phase is clearly separated - we explicitly compute init events and process them. What we don't control at this level is the process of serialization and sending out the events. In the initial version we will ignore this and simply send the initialization done signal as soon as all the init events are in the result channel. In the future, we can estimate it by assuming some constant (or proportional to event size) time for an event and delay delivering the signal by the sum of those across all init events."

@wojtek-t @tkashem I still think we need a timeout on watch InitEvents but it looks promising that a slow client receiving InitEvents cannot hold seats indefinitely.

**Comment by natherz97** [comment]
For making the writer respect the context, would writing the response in chunks and checking if the context has been canceled between each chunk be worth exploring? This isn't perfect because the call to write could still block indefinitely, however, if the number of bytes in our chunk size could be written to the client, it would allow a timeout to enforced for slow (but not hanging) clients. 

I ran some initial tests with writing the response in chunks with a slow client receiving 2000 bytes/s:
```
$ curl -v -k --header "Authorization: Bearer ${TOKEN}" -X GET "${APISERVER}/api/v1/configmaps" --limit-rate 2000
```

Without this patch, the list configmaps request completed in over 1 hour:
```
I0802 00:01:54.412070      10 trace.go:236] Trace[893363742]: "List" accept:*/*,audit-id:6b4b910e-bd90-43c8-ad55-c3496b8a6d3a,client:205.251.233.233,api-group:,api-version:v1,name:,subresource:,namespace:,protocol:HTTP/2.0,resource:configmaps,scope:cluster,url:/api/v1/configmaps,user-agent:curl/8.3.0,verb:LIST (01-Aug-2024 22:47:29.168) (total time: 4465244ms):
Trace[893363742]: ["List(recursive=true) etcd3" audit-id:6b4b910e-bd90-43c8-ad55-c3496b8a6d3a,key:/configmaps,resourceVersion:,resourceVersionMatch:,limit:0,continue: 4465243ms (22:47:29.168)
Trace[893363742]:  ---"fetched all pages from etcd" page-count:23 937ms (22:47:30.106)]
Trace[893363742]: ---"Writing http response done" count:45193 4464305ms (00:01:54.412)
Trace[893363742]: [1h14m25.244005955s] [1h14m25.244005955s] END
E0802 00:01:54.412260      10 timeout.go:142] post-timeout activity - time-elapsed: 1.505983ms, GET "/api/v1/configmaps" result: <nil>
```
With this patch, the list configmaps request completed in 1 minute and 6 seconds:
```
I0801 22:24:56.534337       9 trace.go:236] Trace[123593309]: "List" accept:*/*,audit-id:34a174b3-51e3-4c80-8591-9182088ad8cd,client:205.251.233.233,api-group:,api-version:v1,name:,subresource:,namespace:,protocol:HTTP/2.0,resource:configmaps,scope:cluster,url:/api/v1/configmaps,user-agent:curl/8.3.0,verb:LIST (01-Aug-2024 22:23:49.735) (total time: 66798ms):
Trace[123593309]: ["List(recursive=true) etcd3" audit-id:34a174b3-51e3-4c80-8591-9182088ad8cd,key:/configmaps,resourceVersion:,resourceVersionMatch:,limit:0,continue: 66798ms (22:23:49.735)
Trace[123593309]:  ---"fetched all pages from etcd" page-count:23 845ms (22:23:50.581)]
Trace[123593309]: ---"Writing http response done" count:45193 65952ms (22:24:56.534)
Trace[123593309]: [1m6.798532261s] [1m6.798532261s] END
E0801 22:24:56.534521       9 timeout.go:142] post-timeout activity - time-elapsed: 1.4805ms, GET "/api/v1/configmaps" result: <nil>
```
I did have some concerns about the client perspective in this mode, specifically, we would need to make sure the HTTP request fails with an error to denote that the full response body was not sent. My understanding is that the client would need to detect this through either Transfer-Encoding for HTTP1.1 or through a stream error for HTTP2.

Would love to hear your thoughts on this approach and if we can leverage this in the short-term until https://github.com/kubernetes/kubernetes/pull/114189. @wojtek-t @tkashem @shyamjvs @mengqiy

**Comment by tkashem** [comment]
> would writing the response in chunks and checking if the context has been canceled between each chunk be worth exploring

I think a "API streaming" will help with that cc @p0lyn0mial 


> the call to write could still block indefinitely

Yes, I agree, it's orthogonal to the above, and I believe https://github.com/kubernetes/kubernetes/pull/114189 will address this issue

**Comment by natherz97** [comment]
I was discussing with @tkashem last week and we were trying to figure out why a panic in the golang HTTP handler isn't sufficient to fix this issue. The WithTimeoutForNonLongRunningRequests wrapper will result in a panic if the server-side timeout is hit: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L260-L274
```
func (tw *baseTimeoutWriter) timeout(err *apierrors.StatusError) {
	tw.mu.Lock()
	defer tw.mu.Unlock()

	tw.timedOut = true

	// The timeout writer has not been used by the inner handler.
	// We can safely timeout the HTTP request by sending by a timeout
	// handler
	if !tw.wroteHeader && !tw.hijacked {
		tw.w.WriteHeader(http.StatusGatewayTimeout)
		enc := json.NewEncoder(tw.w)
		enc.Encode(&err.ErrStatus)
	} else {
		// The timeout writer has been used by the inner handler. There is
		// no way to timeout the HTTP request at the point. We have to shutdown
		// the connection for HTTP1 or reset stream for HTTP2.
		//
		// Note from the golang's docs:
		// If ServeHTTP panics, the server (the caller of ServeHTTP) assumes
		// that the effect of the panic was isolated to the active request.
		// It recovers the panic, logs a stack trace to the server error log,
		// and either closes the network connection or sends an HTTP/2
		// RST_STREAM, depending on the HTTP protocol. To abort a handler so
		// the client sees an interrupted response but the server doesn't log
		// an error, panic with the value ErrAbortHandler.
		//
		// We are throwing http.ErrAbortHandler deliberately so that a client is notified and to suppress a not helpful stacktrace in the logs
		panic(http.ErrAbortHandler)
	}
}
```
In the case the timeout occurs while writing the HTTP response to the client we will either:
- HTTP1.1: close the underlying TCP connection and the client will detect it didn't receive the full response from the zero-chunk not being sent sent with chunked transfer encoding or the expected content-length isn't satisifed.
- HTTP2: the current stream will be terminated with an RST_STREAM but the connection stays open.

I ran some tests to determine why this mechanism isn't force closing the connection and the issue is that the tw.mu.Lock() call is blocking in this timeout function above before we can panic. It looks like the go routine handling writing the HTTP response is holding the lock in this function: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L196
```
func (tw *baseTimeoutWriter) Write(p []byte) (int, error) {
	tw.mu.Lock()
	defer tw.mu.Unlock()

	if tw.timedOut {
		return 0, http.ErrHandlerTimeout
	}
	if tw.hijacked {
		return 0, http.ErrHijacked
	}

	if !tw.wroteHeader {
		copyHeaders(tw.w.Header(), tw.handlerHeaders)
		tw.wroteHeader = true
	}
	return tw.w.Write(p)
}
```
I modified the timeout function to not acquire the lock in the panic case and this solves the issue for HTTP1.1 where now we can't exceed the timeout. Here are apiserver logs with the timeout being respected:
```
E0404 22:24:17.166914      11 status.go:87] "Unhandled Error" err="apiserver received an error that is not an metav1.Status: &tls.permanentError{err:(*net.OpError)(0xc00eee8000)}: write tcp 127.0.0.1:443->127.0.0.1:43208: use of closed network connection" logger="UnhandledError"
I0404 22:24:17.170319      11 trace.go:236] Trace[1675545948]: "List" accept:*/*,audit-id:a6fd99ba-b573-4cf8-abb6-e6c4149bc6ab,client:127.0.0.1,api-group:,api-version:v1,name:,subresource:,namespace:,protocol:HTTP/1.1,resource:configmaps,scope:cluster,url:/api/v1/configmaps,user-agent:curl/8.3.0,verb:LIST (04-Apr-2025 22:23:17.164) (total time: 60005ms):
Trace[1675545948]: ["SerializeObject" audit-id:a6fd99ba-b573-4cf8-abb6-e6c4149bc6ab,method:GET,url:/api/v1/configmaps,protocol:HTTP/1.1,mediaType:application/json,encoder:{"encodeGV":"v1","encoder":"{\"name\":\"json\",\"pretty\":\"true\",\"strict\":\"false\",\"yaml\":\"false\"}","name":"versioning"} 59995ms (22:23:17.175)
Trace[1675545948]:  ---"Write call failed" writer:*responsewriter.outerWithCloseNotifyFlushAndHijack,size:31419784,firstWrite:true,err:write tcp 127.0.0.1:443->127.0.0.1:43208: use of closed network connection 59453ms (22:24:17.165)]
Trace[1675545948]: [1m0.005806652s] [1m0.005806652s] END
E0404 22:24:17.170639      11 timeout.go:140] "Post-timeout activity" logger="UnhandledError" timeElapsed="6.258468ms" method="GET" path="/api/v1/configmaps" result=null
```
Here's the HTTP 1.1 curl client getting the timeout after 1 minute:
```
* Closing connection
} [5 bytes data]
* TLSv1.2 (OUT), TLS alert, close notify (256):
} [2 bytes data]
curl: (18) transfer closed with outstanding read data remaining
21468697
```
Unfortunately, this doesn't solve the issue for HTTP2 by removing the lock in the panic case. I can see that the RST_STREAM is sent to the client when we don't block before panic'ing, however, it appears that the mechanism which sends the RST_STREAM frame is blocking on the client receiving it. This is shown with apiserver logs not respecting the timeout and this request taking 5 minutes even after panic'ing:
```
I0404 22:22:24.473713      11 trace.go:236] Trace[870964421]: "List" accept:*/*,audit-id:b97a3bd8-67d4-4c8a-9fcc-e03e6370a394,client:127.0.0.1,api-group:,api-version:v1,name:,subresource:,namespace:,protocol:HTTP/2.0,resource:configmaps,scope:cluster,url:/api/v1/configmaps,user-agent:curl/8.3.0,verb:LIST (04-Apr-2025 22:17:12.807) (total time: 311666ms):
Trace[870964421]: ["SerializeObject" audit-id:b97a3bd8-67d4-4c8a-9fcc-e03e6370a394,method:GET,url:/api/v1/configmaps,protocol:HTTP/2.0,mediaType:application/json,encoder:{"encodeGV":"v1","encoder":"{\"name\":\"json\",\"pretty\":\"true\",\"strict\":\"false\",\"yaml\":\"false\"}","name":"versioning"} 311647ms (22:17:12.826)
Trace[870964421]:  ---"Write call succeeded" writer:responsewriter.outerWithCloseNotifyAndFlush,size:31419784,firstWrite:true 310994ms (22:22:24.473)]
Trace[870964421]: [5m11.666159584s] [5m11.666159584s] END
```
The HTTP2 curl client does receive the RST_STREAM but after 5 minutes (when it receives the full response body):
```
* Connection #0 to host localhost left intact
curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
31419784
```
I'd like to hear if anyone has any thoughts about how we could solve this for HTTP2. One option would be to send a GOAWAY + force connection closure for HTTP2 even if there's other in flight streams (I'm not aware of the apiserver doing this outside of server shutdown or setting the connection close header for HTTP2 for unauthenticated clients or for goaway-chance).

CC: @wojtek-t @sttts
